spatial/curve/hilbert.go

```diff
@@ -0,0 +1,236 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package curve
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ}, where k
+// is the order.
+func (h Hilbert2D) Dims() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (2) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Len() int { return 1 << (2 * h.Order) }
+
+func (h Hilbert2D) rot(n int, v []int, d int) {
+	switch d {
+	case 0:
+		swap{0, 1}.do(n, v)
+	case 3:
+		flip{0, 1}.do(n, v)
+	}
+}
+
+// Curve returns the curve coordinate of v. Curve modifies v.
```

Comment:
If you took the curve and straightened it out, and called one end 'zero' and the other 2ⁿᵏ-1 and added a point for every integer in between, then folded the line back into a Hilbert curve, the 'curve coordinate' would be the number at any given point.

That's not a particularly good description. I haven't been able to come up with a better one.

Reply:
> If we define position 0 as the end of the curve at `(0, ...)` and position 2ⁿᵏ-1 as the other end of the curve, the 'curve coordinate' of a point on the curve is the position along the curve relative to the two ends.

WDYT?

Reply:
Since you are using position as the reference for what a coordinate it, I'd just call it the position and rename the method to "Pos" (similarlly, s/Space(?)D/Coord$1D/g and s/Space/Coord/g), with the comment 
```
// Pos returns the linear position of the spatial coordinate along the curve.
```